Steelseries Rival 600 support

ergor opened the pull request:
~_WIP -  do not merge_~

Added support for changing DPI, polling rate and colors as well as all the color modes (on/off/cycle/breathing). Also added an SVG file (still WIP, nearly complete).

Of course, feedback, criticisms and suggestions are very welcome.

Regarding the driver:
This mouse uses a new DeviceVersion: version 3. It is a frankensteined combo of version 1 and 2. On one hand, the Rival 600 uses the same command codes as mice in version 1; on the other hand, it uses a very similar protocol for colors and effects as version 2 (Rival310 and Sensei310).

Thus, I am piggybacking on _steelseries_write_led_v2()_ for the LED color protocol, because other than the header field indexes they are identical. I modified _construct_cycle_buffer()_ to retrieve field indexes from a _struct steelseries_led_cycle_spec_, instead of using hard-coded indexes. 

I did this in an attempt to reuse code. Should I keep it like that, or should I revert the changes in v2 and have a completely separate v3?

Also, since steelseries_write_led_v2() is modified, could anyone test with a v2 mouse to verify I didn't break anything?

FFY00 commented:
I took a glance at the code and it seemed fine. I'll properly review this once I get more time, if I am taking to long please @ me. Seriously, don't be afraid to mention me.

Regarding your questions, IMO if both protocols are working it would be better to have them sharing the code.

FFY00 commented:
@bentiss can you have a look at the SVG?

FFY00 commented:
I also forgot to mention, please don't use merges, rebase instead.

ergor commented:
@FFY00 
Rebased and finished up the SVG. Please verify i didn't mess up the git history. I had to squash my commits with the merge commit before rebasing.

FFY00 commented:
The git history looks fine to me. I will test and review this today if I have the opportunity.

Hjdskes commented:
SVG looks great!

FFY00 commented on src/driver-steelseries.c:
```diff
@@ -597,50 +635,92 @@ create_cycle(struct steelseries_led_cycle *cycle)
+		default:
+			return -1;
+	}
+}
```
@phomes isn't there a better way to do this? I feel this should be based on the protocol version, not the device IDs.

ergor replied:
Yes, that does make more sense

FFY00 commented:
The driver seems to be working on my mouse.

If you want you can test the read commands from the protocol 2. I haven't implemented yet but it should be good to know if they're compatible with protocol 3.
https://github.com/FFY00/rival310-re

ergor commented:
@FFY00 I tried the read commands, however it doesn't seem like it did anything:
[rival600-read-attempts.txt](https://github.com/libratbag/libratbag/files/2398151/rival600-read-attempts.txt)
Are there any known read commands for v1 that I should try?

FFY00 commented:
I can't see the packet dumps there.

This is the easier way of testing commands I've been able to come up with:
Run `hexdump /dev/hidraw0` (should be 0 if you don't have any other HID device connected but it might be different on your system) and keep it opened to read from the mouse. Then you can send a command using something like this https://gist.github.com/FFY00/faa3065a0fd39681143714c6272d90af (again, should be `hidraw1` but it might be other device).

ergor commented:
The data sent is in the `URB setup->Data Fragment` field.
In the previously attached file, you can see the responses from the mouse say:
`Data: not present ('>')`

I tried on hidraw0-2, but got nothing

dmesg output:
```
[109841.528964] hid-generic 0003:1038:1724.0014: hiddev0,hidraw0: USB HID v1.11 Device [SteelSeries SteelSeries Rival 600] on usb-0000:00:14.0-2/input0
[109841.531710] input: SteelSeries SteelSeries Rival 600 as /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.1/0003:1038:1724.0015/input/input47
[109841.532014] hid-generic 0003:1038:1724.0015: input,hidraw1: USB HID v1.11 Mouse [SteelSeries SteelSeries Rival 600] on usb-0000:00:14.0-2/input1
[109841.533891] input: SteelSeries SteelSeries Rival 600 as /devices/pci0000:00/0000:00:14.0/usb1/1-2/1-2:1.2/0003:1038:1724.0016/input/input48
[109841.585643] hid-generic 0003:1038:1724.0016: input,hidraw2: USB HID v1.11 Keyboard [SteelSeries SteelSeries Rival 600] on usb-0000:00:14.0-2/input2
```

writing:
```
$ echo -en "\x90\x00\x00\x00" > /dev/hidraw0
$ echo -en "\x90\x00\x00\x00" > /dev/hidraw1
$ echo -en "\x90\x00\x00\x00" > /dev/hidraw2
```

reading:
```
$ hexdump /dev/hidraw0
^C
$ hexdump /dev/hidraw1
^C
$ hexdump /dev/hidraw2
^C
```

Any ideas?

FFY00 commented:
Huh, maybe it wasn't implemented in that mouse.

ergor commented:
I don't have my windows computer with me at the moment, but I'll try to do some more sniffing when I get it, see if I can fish out the read commands from the official tool

FFY00 commented:
Yeah. If you can post a packet dump of the driver being started here for me to see :smile:.

ergor commented:
Will do :+1: 

BTW, refactored the led cycle spec to be based on device version instead of usb id:
- removed create_cycle_spec()
- the function formerly known as steelseries_write_led_v2() is now steelseries_write_led_cycle()
- now steelseries_write_led_v2() and steelseries_write_led_v3() are wrapper functions that creates their respective cycle specs and passes it to steelseries_write_led_cycle()

FFY00 commented:
Looks good. Not sure if we should call the shared led write function `steelseries_write_led()`, perhaps we should use a less ambiguous name. @whot can you comment on that?

Other thin, I just noticed https://github.com/libratbag/libratbag/blob/0c86709cde0a139441df056a1da3d62390fc9968/src/driver-steelseries.c#L413. What happens when `0x42` isn't passed into the buffer? In my Rival 310 it doesn't seem to be mandatory.

ergor commented:
[pcap (wireshark v. >=1.8)](https://www.netb.st/projects/rival600/protocol/downloads/)
Inserting the mouse generated packets up to no. 108. 
No. 109 and to out are the ones generated when starting Steelseries Engine.

So far I got this:

```
COMMAND     DESCRIPTION                 RESPONSE
0x10        Read firmware version       2 bytes (LE; MSB:major, LSB: minor)
0x16        Read active DPI setting?    1 byte?
```

Regarding the `0x42`:
I could have sworn it didn't work for me without it during previous tests. But when I tested it again to confirm, it worked every time even without it. Weird... Guess it should be changed to _unsure_ for this one as well

FFY00 commented:
> [pcap (wireshark v. >=1.8)](https://www.netb.st/projects/rival600/protocol/downloads/)
> Inserting the mouse generated packets up to no. 108.
> No. 109 and out are the ones generated when starting Steelseries Engine.
> 
> So far I got this:
> 
> ```
> COMMAND     DESCRIPTION                 RESPONSE
> 0x10        Read firmware version       2 bytes (LE; MSB:major, LSB: minor)
> 0x16        Read active DPI setting?    1 byte?
> ```

Implement them :smile:.

> EDIT: Just to add, there is one more command in the capture file that triggers a read, but I couldn't make out what it means
> 
> Regarding the `0x42`:
> I could have sworn it didn't work for me without it during previous tests. But when I tested it again to confirm, it worked every time even without it. Weird... Guess it should be changed to _unsure_ for this one as well

Sure.

ergor commented:
@FFY00 All done now. Added support for button config and implemented all the relevant read functions the mouse supports (sadly there's no read current config command).

FFY00 commented on src/driver-steelseries.c:
```diff
@@ -189,8 +215,11 @@ steelseries_read_settings(struct ratbag_device *device)
+	else if (device_version == 2)
+		buf[0] = STEELSERIES_ID_SETTTINGS;
+	else if (device_version == 3)
+		buf[0] = STEELSERIES_ID_GET_DPI_SHORT;
```
I think this should follow the current naming scheme or vice-versa.

FFY00 commented on src/driver-steelseries.c:
```diff
@@ -157,8 +180,11 @@ steelseries_get_firmware_version(struct ratbag_device *device)
+	else if (device_version == 2)
+		buf[0] = STEELSERIES_ID_FIRMWARE;
+	else if (device_version == 3)
+		buf[0] = STEELSERIES_ID_FIRMWARE_SHORT;
```
Why short? Should this be `STEELSERIES_ID_FIRMWARE_PROTOCOL3` or something like that?

FFY00 commented on src/driver-steelseries.c:
```diff
@@ -375,12 +413,18 @@ steelseries_write_dpi(struct ratbag_resolution *resolution)
 		buf[6] = 0x42; /* not sure if needed */
+	} else {
+		buf_len = STEELSERIES_REPORT_SIZE;
+		buf[0] = STEELSERIES_ID_DPI_SHORT;
```
Again, shouldn't be `..._SHORT` IIRC.

ergor commented:
Added own defines. 
Also, was it intended behavior to write to the mouse regardless of wether the device version was implemented?
I added returning of `-ENOTSUP` for unimplemented device versions before doing any writing. Is this OK?

FFY00 commented:
Sorry for the late response! I have been a bit busy and I totally forgot. You could have pinged me :). @phomes can you take a look at this?

FFY00 commented:
Perhaps @whot could look at this since @phomes is inactive.

@ergor I am really sorry for this huge wait time, we all have been busy.

ergor commented:
@FFY00 No worries! It was on me to ping more frequently anyway 🙂

whot commented on src/driver-steelseries.c:
```diff
@@ -200,19 +234,28 @@ steelseries_read_settings(struct ratbag_device *device)
-		}
+				resolution->dpi_x = 100 * (1 +  buf[2 + resolution->index*2]);
+				resolution->dpi_y = resolution->dpi_x;
+			}
```
i think there's some indentation issue here?

whot commented:
looks good on a glance, can you double-check the indentation issue please and squash them all together? I'll hid the merge-button as soon as you force-pushed to the branch then :) Oh, and ping me because github doesn't send notifications on a force-push.

ergor commented:
@whot The indentation seems to be fine (see my response comment). Also squashed and force pushed, but the build started failing CircleCI. Tried to rebase onto upstream master, but no luck:
```
#!/bin/bash -eo pipefail
python3 ~/libratbag/tools/check_scan_build.py ~/libratbag/build/meson-logs/scanbuild

2 bugs are ignored:
    warning: Potential leak of memory pointed to by 'data' at /root/libratbag/src/hidpp20.c:2365
    warning: Potential leak of memory pointed to by 'data' at /root/libratbag/src/libratbag-data.c:412

ERROR: 8 bugs found:
  * WARNING: Assigned value is garbage or undefined at /root/libratbag/src/hidpp10.c:1520
  * WARNING: Assigned value is garbage or undefined at /root/libratbag/src/hidpp10.c:1512
  * WARNING: Access to field '__parent_and_color' results in a dereference of a null pointer (loaded from variable 'n') at /root/libratbag/src/shared-rbtree.h:83
  * WARNING: Dereference of null pointer at /root/libratbag/src/hidpp10.c:1171
  * WARNING: Access to field '__parent_and_color' results in a dereference of a null pointer (loaded from variable 'n') at /root/libratbag/src/shared-rbtree.h:83
  * WARNING: Assigned value is garbage or undefined at /root/libratbag/src/hidpp10.c:1500
  * WARNING: Potential leak of memory pointed to by 'macro' at /root/libratbag/src/hidpp20.c:2029
  * WARNING: 1st function call argument is an uninitialized value at /root/libratbag/src/hidpp10.c:1462
Exited with code 1
```

FFY00 commented:
Shouldn't be. Also, you're only failing on fedora, so @whot might have more information.

whot commented:
Thanks, I fixed the CI issues for now with #560 until we have time to go through #556 and address them accordingly
